colossalai/checkpoint_io/hybrid_parallel_checkpoint_io.py

```diff
@@ -0,0 +1,339 @@
+import copy
+import logging
+import os
+from pathlib import Path
+from shutil import rmtree
+from typing import Any, Callable, Iterator, List, Optional, OrderedDict, Tuple, Union
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+from torch.optim import Optimizer
+from torch.optim.lr_scheduler import _LRScheduler as LRScheduler
+
+from colossalai.cluster import ProcessGroupMesh
+from colossalai.tensor.d_tensor import (
+    is_customized_distributed_tensor,
+    is_distributed_tensor,
+    to_global,
+    to_global_for_customized_distributed_tensor,
+)
+
+from .general_checkpoint_io import GeneralCheckpointIO
+from .index_file import CheckpointIndexFile
+from .utils import (
+    calculate_tensor_size,
+    get_model_base_filenames,
+    get_optimizer_base_filenames,
+    get_shard_filename,
+    is_safetensors_available,
+    load_shard_state_dict,
+    load_state_dict_into_model,
+    save_param_groups,
+    save_state_dict,
+    save_state_dict_shards,
+)
+
+try:
+    from torch.nn.modules.module import _EXTRA_STATE_KEY_SUFFIX, _IncompatibleKeys
+except ImportError:
+    _EXTRA_STATE_KEY_SUFFIX = '_extra_state'
+
+DP_AXIS, PP_AXIS, TP_AXIS = 0, 1, 2
+
+
+class HypridParallelCheckpointIO(GeneralCheckpointIO):
+    """
+    CheckpointIO for Hybrid Parallel Training.
+
+    Args:
+        pg_mesh (ProcessGroupMesh): Process group mesh containing information of process groups along different dimensions.
+    """
+
+    def __init__(self, pg_mesh: ProcessGroupMesh) -> None:
+        super().__init__()
+        self.dp_group = pg_mesh.get_group_along_axis(DP_AXIS)
+        self.pp_group = pg_mesh.get_group_along_axis(PP_AXIS)
+        self.tp_group = pg_mesh.get_group_along_axis(TP_AXIS)
+        self.dp_rank = dist.get_rank(self.dp_group)
+        self.tp_rank = dist.get_rank(self.tp_group)
+        self.pp_rank = dist.get_rank(self.pp_group)
+        self.dp_size, self.pp_size, self.tp_size = pg_mesh.shape
+
+    @staticmethod
+    def _model_sharder(model: nn.Module,
+                       prefix: str = '',
+                       keep_vars: bool = False,
+                       size_per_shard: int = 1024) -> Iterator[Tuple[OrderedDict, int]]:
+        # An internel method that breaks state_dict of model into shards within limited size.
+
+        state_dict_sharder = _StateDictSharder(size_per_shard)
+
+        # Save parameters.
+        for name, param in model.named_parameters():
+            if param is None:
+                continue
+            # Gather tensor pieces when using tensor parallel.
+            param_ = param if keep_vars else param.detach()
+            if is_distributed_tensor(param_):
+                param_ = to_global(param_)
```

Comment:
It seems that this part is similar to `ParallelModule`. Can you extract this part to a common util function?

Reply:
modified